bridges/snowbridge/pallets/system-v2/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023 Snowfork <hello@snowfork.com>
+//! Governance API for controlling the Ethereum side of the bridge
+//!
+//! # Extrinsics
+//!
+//! ## Agents
+//!
+//! Agents are smart contracts on Ethereum that act as proxies for consensus systems on Polkadot
+//! networks.
+//!
+//! * [`Call::create_agent`]: Create agent for any kind of sovereign location on Polkadot network,
+//!   can be a sibling parachain, pallet or smart contract or signed account in that parachain, etc
+
+//! ## Polkadot-native tokens on Ethereum
+//!
+//! Tokens deposited on AssetHub pallet can be bridged to Ethereum as wrapped ERC20 tokens. As a
+//! prerequisite, the token should be registered first.
+//!
+//! * [`Call::register_token`]: Register a token location as a wrapped ERC20 contract on Ethereum.
+#![cfg_attr(not(feature = "std"), no_std)]
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+pub mod api;
+pub mod weights;
+pub use weights::*;
+
+use frame_support::{pallet_prelude::*, traits::EnsureOrigin};
+use frame_system::pallet_prelude::*;
+use snowbridge_core::{AgentId, AssetMetadata, TokenId, TokenIdOf};
+use snowbridge_outbound_queue_primitives::{
+	v2::{Command, Message, SendMessage},
+	SendError,
+};
+use sp_core::H256;
+use sp_runtime::traits::MaybeEquivalence;
+use sp_std::prelude::*;
+use xcm::prelude::*;
+use xcm_executor::traits::ConvertLocation;
+
+#[cfg(feature = "runtime-benchmarks")]
+use frame_support::traits::OriginTrait;
+
+pub use pallet::*;
+
+pub type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+
+pub fn agent_id_of<T: Config>(location: &Location) -> Result<H256, DispatchError> {
+	T::AgentIdOf::convert_location(location).ok_or(Error::<T>::LocationConversionFailed.into())
+}
+
+#[cfg(feature = "runtime-benchmarks")]
+pub trait BenchmarkHelper<O>
+where
+	O: OriginTrait,
+{
+	fn make_xcm_origin(location: Location) -> O;
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::dispatch::RawOrigin;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Send messages to Ethereum
+		type OutboundQueue: SendMessage;
+
+		/// Origin check for XCM locations that can create agents
+		type SiblingOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Location>;
+
+		/// Converts Location to AgentId
+		type AgentIdOf: ConvertLocation<AgentId>;
+
+		type WeightInfo: WeightInfo;
```

Comment:
@yrong move to end, please

Reply:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/381c730d33d130c37bafba013aef0139459b3aff